### ISSUE TITLE
Some tips for multithread debugging?

### ISSUE THREAD

**Issue by nullbus** [Microsoft/ChakraCore#5463]
Hello, I am working on some porting from V8 implemented [Unreal Engine Javascript plugin](https://github.com/ncsoft/Unreal.js-core) to [ChakraCore implementation](https://github.com/ninemcom/Unreal.js-core).

Its primary objective is supporting mobile devices(iOS, Android) and I got success from most devices. But I got reports about thread hanging issue for some old Samsung Android devices(like Galaxy S6). This problem exists even if I set flag `JsRuntimeAttributeDisableBackgroundWork` to disable threaded work. Unfortunately I couldn't use native debugger to that devices due to configuration.

To investigate the issue, I rebuilt it with `--trace` and some extra multithread traces turned on, but it always ends with waiting trace message at random timing. The trace log file is very huge I cannot just paste all contents to here. But here is last few lines:
```
07-09 16:22:24.680 26488 26591 E chakracore-log: [LCS-UN 0xcfd19670] ==> OK
07-09 16:22:24.680 26488 26591 E chakracore-log: ] VirtualAlloc_ /Users/9m/chakracore-build/ChakraCore/pal/src/map/virtual.cpp:1693
07-09 16:22:24.680 26488 26591 E chakracore-log: VirtualAlloc returning 0xae721000
07-09 16:22:24.680 26488 26591 E chakracore-log:
07-09 16:22:24.681 26488 26510 E chakracore-log: ] GetTickCount /Users/9m/chakracore-build/ChakraCore/pal/src/misc/time.cpp:185
07-09 16:22:24.681 26488 26510 E chakracore-log: GetTickCount ()
07-09 16:22:24.681 26488 26510 E chakracore-log: ] GetTickCount /Users/9m/chakracore-build/ChakraCore/pal/src/misc/time.cpp:190
07-09 16:22:24.681 26488 26510 E chakracore-log: GetTickCount returns DWORD 437670471
07-09 16:22:24.681 26488 26510 E chakracore-log: ] VirtualAlloc_ /Users/9m/chakracore-build/ChakraCore/pal/src/map/virtual.cpp:1625
07-09 16:22:24.681 26488 26510 E chakracore-log: VirtualAlloc(lpAddress=0xab950000, dwSize=77824, flAllocationType=0x1000,           flProtect=0x4)
07-09 16:22:24.681 26488 26510 E chakracore-log: ] InternalEnterCriticalSection /Users/9m/chakracore-build/ChakraCore/pal/src/sync/cs.cpp:773
07-09 16:22:24.681 26488 26510 E chakracore-log: [ECS 0xcfd19670] Switching from {0, 0, 0} to {0, 0, 1} ==>
07-09 16:22:24.681 26488 26510 E chakracore-log: ] InternalEnterCriticalSection /Users/9m/chakracore-build/ChakraCore/pal/src/sync/cs.cpp:785
07-09 16:22:24.681 26488 26510 E chakracore-log: [ECS 0xcfd19670] ==> OK LockCount={0, 0, 1} lVal={0, 0, 0}
07-09 16:22:24.681 26488 26510 E chakracore-log: ] VIRTUALFindRegionInformation /Users/9m/chakracore-build/ChakraCore/pal/src/map/virtual.cpp:536
07-09 16:22:24.681 26488 26510 E chakracore-log: VIRTUALFindRegionInformation( 0xab950000 )
07-09 16:22:24.681 26488 26510 E chakracore-log: ] VIRTUALCommitMemory /Users/9m/chakracore-build/ChakraCore/pal/src/map/virtual.cpp:1117
07-09 16:22:24.681 26488 26510 E chakracore-log: Committing the memory now..
07-09 16:22:24.697 26488 26591 E chakracore-log: ] VirtualAlloc_ /Users/9m/chakracore-build/ChakraCore/pal/src/map/virtual.cpp:1625
07-09 16:22:24.700 26488 26591 E chakracore-log: VirtualAlloc(lpAddress=0xae722000, dwSize=4096, flAllocationType=0x1000,           flProtect=0x4)
07-09 16:22:24.700 26488 26591 E chakracore-log: ] PALCS_WaitOnCS /Users/9m/chakracore-build/ChakraCore/pal/src/sync/cs.cpp:1202
07-09 16:22:24.700 26488 26591 E chakracore-log: [WCS 0xcfd19670] Switching from {0, 0, 1} to {1, 0, 1} ==>
07-09 16:22:24.700 26488 26591 E chakracore-log: ] PALCS_WaitOnCS /Users/9m/chakracore-build/ChakraCore/pal/src/sync/cs.cpp:1208
07-09 16:22:24.700 26488 26591 E chakracore-log: [WCS 0xcfd19670] ==> OK
07-09 16:22:24.700 26488 26591 E chakracore-log: ] PALCS_DoActualWait /Users/9m/chakracore-build/ChakraCore/pal/src/sync/cs.cpp:1237
07-09 16:22:24.700 26488 26591 E chakracore-log: Trying to go to sleep [CS=0xcfd19670]
07-09 16:22:24.700 26488 26591 E chakracore-log: ] PALCS_DoActualWait /Users/9m/chakracore-build/ChakraCore/pal/src/sync/cs.cpp:1247
07-09 16:22:24.700 26488 26591 E chakracore-log: Actually Going to sleep [CS=0xcfd19670]
```
It is one of hanging points I have got.

I know ChakraCore does not support these mobile devices officially yet but any tips or helps are very appreciated!

**Comment by nullbus** [comment]
I didn't do the whole porting jobs. I just found somebody did and do some minor fixes to get build and work properly with armv7.
For Android: [gashtio's port](https://github.com/gashtio/ChakraCore/tree/feature/android)
For iOS: [chakrashim part of node-chakracore](https://github.com/nodejs/node-chakracore/tree/master/deps/chakrashim)

I am possible to get memory dump of hanged application with this [tool](https://github.com/friendlyJLee/pmdump). So do you want the whole content or specific part of the memory?

**Comment by nullbus** [comment]
@digitalinfinity I shared some log with memory dump from my [drive](https://drive.google.com/open?id=16WfXwReyBLkoUzUmgDMlXKCSfcqFduP4). You can view the content of dump using the [parser script](https://github.com/friendlyJLee/pmdump#pmdump_parserpy). I will add more files as issue reproduced again.

**Comment by nullbus** [comment]
For followup:

I had a chance to get a device that native debugger is attachable. So I narrowed down the problem to this line https://github.com/Microsoft/ChakraCore/blob/v1.8.0/pal/src/map/virtual.cpp#L1188 
Actually a thread hanged during `memset`, I think some kind of memory access violation occured.

The address in `StartBoundary` is `mprotect`ed before calling `memset`. I changed `MMAP_DOESNOT_ALLOW_REMAP` flag to 0 to call `mmap` instead of `mprotect` and problem solved. I still do not know why only few Samsung devices have this issue. I close this issue but keep watch on it.
